Add Python installation guide and Cotainr documentation

Chroxvi opened the pull request:
The main part of this PR is:
- Adding a new page on "Installing Python packages" on LUMI, giving an overview of recommended and discouraged options for installing Python packages on LUMI. This is includes a condensed version of the content in the stale #24. The intention is to close #24 as superseded by this PR.
- Updating the existing documentation on pulling and building Singularity/Apptainer containers on LUMI to include a description of the use of [cotainr](https://cotainr.readthedocs.io/en/stable/) to build containers on LUMI (primarily for Conda/pip environments).
- Updating the "container wrapper" documentation to reflect the more general purpose nature of the tool - not just a Conda/pip wrapper. This includes a tentative change of its name to "Tykky installation wrapper" - please suggest a better name, if you have one! See #110 for more discussion about the motivation for this change.

Additionally, this PR also includes:
- Minor changes and clean-ups to various pages throughout the documentation mentioning do's and don'ts related to Conda, pip, or Python. These now link to the new "Installing Python packages" page.
- A fix for #60 

Open questions:
- [x] Is "Tykky installation wrapper" the right new name for what is currently known as the "LUMI container wrapper".
- [x] If we rename the "LUMI container wrapper" in the docs, how do we handle the transition of the [lumi-container-wrapper](https://lumi-supercomputer.github.io/LUMI-EasyBuild-docs/l/lumi-container-wrapper/) module to a new name?

Chroxvi committed:
Added description of cotainr to singularity docs.

Chroxvi committed:
Added description of ways to "install" Python on LUMI.

Chroxvi committed:
Minor updates and formatting changes to cotainr and Python docs.

Chroxvi committed:
Updated container wrapper page to make the purpose of the tool more clear.

Chroxvi committed:
A few minor fixes to Python and Singularity docs.

Chroxvi committed:
Updated all of the docs to consistently refer to Python install guide.

hdrei commented on docs/software/packages/pytorch.md:
```diff
@@ -1,46 +1,51 @@
+[containers]: ./singularity.md
+[python-install]: ./python.md
```
```suggestion
[python-install]: ./installing/python.md
```

hdrei replied:
Not sure if this is the way to correct the link. But the previous one didn't work at least

hdrei commented on docs/software/packages/pytorch.md:
```diff
@@ -1,46 +1,51 @@
-PyTorch can be run within containers as well. In particular, containers from the images provided by [AMD on DockerHub](https://hub.docker.com/u/rocm).
-Those images are updated frequently and make it possible to try PyTorch with recent ROCm versions. 
-Another point in favor of using containers, is that PyTorch's installation directory can be quite large both in terms of storage size and number of files.
+PyTorch can be installed by following the [official instructions](https://pytorch.org/get-started/locally/) for installing a ROCm compatible PyTorch via pip. Please consults the [Python packages installation guide][python-install] for an overview of recommended ways to manage pip installations on LUMI. Alternatively, the container images provided by [AMD on DockerHub](https://hub.docker.com/u/roc ...
```
```suggestion
PyTorch can be installed by following the [official instructions](https://pytorch.org/get-started/locally/) for installing a ROCm compatible PyTorch via pip. Please consult the [Python packages installation guide][python-install] for an overview of recommended ways to manage pip installations on LUMI. Alternatively, the container images provided by [AMD on DockerHub](https://hub.docker.com/u/rocm) may be used on LUMI.
```

hdrei commented on docs/software/packages/pytorch.md:
```diff
@@ -1,46 +1,51 @@
+[containers]: ./singularity.md
```
```suggestion
[containers]: ./containers/singularity.md
```

hdrei replied:
Not sure if this fixes the link. But the previous link was broken at least

hdrei commented on docs/software/packages/pytorch.md:
```diff
@@ -1,46 +1,51 @@
 ## Multi-GPU training
 
-The communication between LUMI's GPUs during training with Pytorch is done via [RCCL](https://github.com/ROCmSoftwarePlatform/rccl), which is a library of  collective communication routines for GPUs. RCCL works out of the box on LUMI's, however, a special plugin is required so it can take advantage of the [Slingshot interconnect](https://www.hpe.com/emea_europe/en/compute/hpc/slingshot-interconne ...
+The communication between LUMI's GPUs during training with Pytorch is done via [RCCL](https://github.com/ROCmSoftwarePlatform/rccl), which is a library of  collective communication routines for AMD GPUs. RCCL works out of the box on LUMI, however, a special plugin is required so it can take advantage of the [Slingshot 11 interconnect](https://www.hpe.com/emea_europe/en/compute/hpc/slingshot-inter ...
```
Is there a reason why the `Slingshot 11 interconnect` link is to HPE docs and not to the page in our docs? Maybe this could link to our docs, and the Network and interconnect page in our docs to the HPE docs?

Chroxvi replied:
Maybe the "Network and interconnect" page didn't exist when this was written?

I like your suggestion, so I will include it, though it probably belongs in another PR...

hdrei commented on docs/software/index.md:
```diff
@@ -12,25 +14,28 @@
 [software-library]: https://lumi-supercomputer.github.io/LUMI-EasyBuild-docs
 
```
```suggestion
[install-policy]: ./policy.md

```

hdrei commented on docs/software/index.md:
```diff
@@ -12,25 +14,28 @@
+looking for information on how to develop your own software on LUMI, consult
+the [developing section][developing-overview] instead. Please note that the
+LUMI User Support Team can only offer limited help with installing scientific
+software. This is further explained on the [Install policy](/software/policy/)
```
```suggestion
software. This is further explained on the [Install policy][install-policy]
```

hdrei replied:
The link was broken and trying to fix it...

hdrei commented on docs/software/containers/singularity.md:
```diff
@@ -58,42 +62,116 @@ where the command was run. Once the image has been pulled, the container can be
+reasons. Thus, in order to build our own Singularity/Apptainer container for
+LUMI, you have two options:
+
+1. Use [cotainr][cotainr] to build containers on LUMI (only for certain [use
```
```suggestion
**1. Use [cotainr][cotainr-chapter] tool to build containers on LUMI (only for certain use cases).**
```

hdrei commented on docs/software/containers/singularity.md:
```diff
@@ -58,42 +62,116 @@ where the command was run. Once the image has been pulled, the container can be
+LUMI, you have two options:
+
+1. Use [cotainr][cotainr] to build containers on LUMI (only for certain [use
+   cases][cotainr-usecases]).
```
```suggestion
```

hdrei commented on docs/software/containers/singularity.md:
```diff
@@ -58,42 +62,116 @@ where the command was run. Once the image has been pulled, the container can be
+
+1. Use [cotainr][cotainr] to build containers on LUMI (only for certain [use
+   cases][cotainr-usecases]).
+2. Build your own container on your local hardware, e.g. your laptop.
```
```suggestion
**2. [Build your own container][building-local] on your local hardware, e.g. your laptop.**
```

hdrei commented on docs/software/containers/singularity.md:
```diff
@@ -1,16 +1,29 @@
+[python-packages]: ../installing/python.md
 [spack]: ../../software/installing/spack.md
-[copying-files]: ../../firststeps/movingdata.md
 
```
```suggestion
[cotainr-chapter]: ./singularity.md#building-containers-using-cotainr
[building-local]: ./singularity.md#building-containers-on-local-hardware


```

hdrei commented on docs/software/containers/singularity.md:
```diff
@@ -58,42 +62,116 @@ where the command was run. Once the image has been pulled, the container can be
+   cases][cotainr-usecases]).
+2. Build your own container on your local hardware, e.g. your laptop.
+
+### Building containers using cotainr
```
```suggestion
### Building containers using cotainr tool
```

hdrei commented on docs/software/containers/singularity.md:
```diff
@@ -58,42 +62,116 @@ where the command was run. Once the image has been pulled, the container can be
+like any other [container job][container-jobs] on LUMI. It automatically
+activates the installed Conda environment when you run it. See the [cotainr
+Conda environment docs][cotainr-conda-env] and the [cotainr LUMI
+examples][cotainr-lumi-examples] for more details.
```
This link gives a '404 page not found'

Chroxvi replied:
I fixed the link, which was missing an "I" in the end. The correct link is: https://github.com/DeiC-HPC/cotainr/tree/main/examples/LUMI

hdrei commented on docs/software/containers/singularity.md:
```diff
@@ -58,42 +62,116 @@ where the command was run. Once the image has been pulled, the container can be
+Cray-MPICH found on LUMI. That MPICH will be used to compile the [OSU
+microbenchmarks][osu-benchmark]. Finally, the OSU point to point bandwidth test
+is set as the "runscript" of the image.
 
```
This example needs to be updated! Maybe not the task of this PR, but it doesn't work any more. Should it be a separate issue of should we/someone try to fix it?

Chroxvi replied:
I would prefer to keep such an update in a separate PR as it doesn't really have anything to do with either Python or cotainr. So yes, I think we should create an issue detailing what doesn't work. Then someone should fix it in a separate PR. Also this reminds me that we need to determine a strategy for keeping such examples in the docs up-to-date...

hdrei replied:
Ok, I made a note about this

hdrei commented on docs/software/installing/python.md:
```diff
@@ -0,0 +1,137 @@
+A Python installation usually consists of the Python interpreter, the Python
+standard library and one or more third party Python packages. Such Python
+packages may include both compiled code and a lot of so-called Python modules,
+i.e. of a lot of small files containing Python code, e.g. a typical
```
```suggestion
i.e. a lot of small files containing Python code. A typical
```

hdrei commented on docs/software/installing/python.md:
```diff
@@ -0,0 +1,137 @@
+  build a container based on a [Conda environment][conda-env] file for use on
+  LUMI.
+
+### Conditionally recommended
```
```suggestion
### Recommended in certain use cases
```

hdrei replied:
"Conditionally" sounds like that "these might not be recommended" and a reader might skip them and not notice that they are actually easy options for certain use cases

hdrei replied:
In general I think that it might be better to not put the headings "recommended" and "conditionally recommended" (or something), but rather list the different methods as headings. That way they would also appear in the listing on the upper right corner of the page. I'm not sure how to formulate this change, though, at least right away

hdrei commented on docs/software/installing/python.md:
```diff
@@ -0,0 +1,137 @@
+  detailed discussion of when this approach may be preferred over using a
+  container directly.
+
+If you are used to managed software stacks on local HPC systems, and your local
+organization provides a similar software stack on LUMI, you may prefer to use a
+Python installation from such a local organization software stack:
+
+- **Use the CSC software stack**: [CSC provides a software
+  stack][csc-software-stack] on LUMI, similar to the software stacks provided
+  on the Finish HPC systems, which contains some Python packages. Please note
+  that this software stack is only supported by CSC, not the LUMI User Support
+  Team (LUST).
```
```suggestion
```

hdrei replied:
I don't see the point listing this here. Local software collections should be talked about in the overview page of this section, as other software and/or installation methods. Or am I missing a point here why this was added...

Chroxvi replied:
This is mainly to point out that if you are used to e.g. `module load pytorch` on one of the Finish HPC systems, you have a similar option on LUMI, i.e. you might not need to install anything yourself. I will try to make this point more clear.

hdrei commented on docs/software/installing/python.md:
```diff
@@ -0,0 +1,137 @@
+Finally, for the sake of completeness, you can technically also install Python
+packages directly to the file systems using one of the methods listed below.
+However, as already discussed previously on this page, **these methods are
+strongly discouraged**.
```
```suggestion
**We strongly discourage using these methods on LUMI**.
```

hdrei replied:
I think it shouldn't say something as: "Well, this can be done, but we advice against it" type of thing. It doesn't sound serious enough at least to me. If I would be an inpatient person not willing to learn the methods described above, I would just do the "not recommended" way, if it sounds it is not that seriously discouraged

hdrei commented on docs/software/installing/container-wrapper.md:
```diff
@@ -1,43 +1,81 @@
-subset is visible during execution and installation. This means that it's
-possible to wrap installations using e.g. mpi4py while still relying on the
-host provided MPI installation.
+The [Tykky][tykky-github] installation wrapper (formerly the "LUMI container
```
```suggestion
The [Tykky][tykky-github] installation wrapper (formerly called the "LUMI container
```

hdrei commented on docs/software/index.md:
```diff
@@ -12,25 +14,28 @@
 
 ---
-On this page you find guidance on ways to install additional software on LUMI. 
+On this page you find guidance on ways to install additional software on LUMI.
```
```suggestion
On this page you find information of the pre-installed software on LUMI, and guidance on ways to install additional software yourself.
```

hdrei commented on docs/software/index.md:
```diff
@@ -12,25 +14,28 @@
+    based approach must be used. See the [Python packages installation guide
+    ][python-install] for en overview of options for installing Conda/pip
+    environments.
 
```
```suggestion

## Pre-installed software on LUMI
- **Software already installed in the LUMI Software Library**: The [LUMI
software library](https://lumi-supercomputer.github.io/LUMI-EasyBuild-docs/)
contains an overview with all software that is installed in the software stacks
available through the LUMI modules and the CrayEnv module (apart from the
documentation of the HPE Cray Programming Environment). For software installed
on the system the first channel to get help on a specific module is through the
`module help` command, but the LUMI Software Library pages sometimes contain
more information on how to run the software, or more information about specific
options that were chosen when installing the software on LUMI.
- **Local software collections**: There are also software available on LUMI installed by local LUMI organizations. These are available for all LUMI users and are updated and supported by the local LUMI organizations themselves, not by LUST.
    * [CSC installed software collection](https://docs.lumi-supercomputer.eu/software/local/csc/)

```

hdrei replied:
Would this suit here? The text would need to be formulated a bit more, but I think it might be a good idea to list first the options where a user doesn't need to install anything, and how to find the pre-installes software

hdrei commented on docs/software/index.md:
```diff
@@ -58,51 +63,38 @@ The three options are:
+`module help` command, but the LUMI Software Library pages sometimes contain
+more information on how to run the software, or more information about specific
+options that were chosen when installing the software on LUMI.
+- **Use local software collections**: There are also software available on LUMI installed by local LUMI organizations. These are available for all LUMI users and are updated and supported by the local LUMI organizations themselves, not by LUST.
```
```suggestion
As an alternative to installing software yourself, you may:

- **Use an Apptainer/Singularity container**: You can bring a
[Singularity/Apptainer software container][singularity-container] and run it
using the [Singularity container runtime][singularity-jobs] provided by LUMI.

```

hdrei commented on docs/software/index.md:
```diff
@@ -58,51 +63,38 @@ The three options are:
+options that were chosen when installing the software on LUMI.
+- **Use local software collections**: There are also software available on LUMI installed by local LUMI organizations. These are available for all LUMI users and are updated and supported by the local LUMI organizations themselves, not by LUST.
+
+    * [CSC installed software collection](https://docs.lumi-supercomputer.eu/software/local/csc/)
```
```suggestion

```

hdrei commented on docs/software/index.md:
```diff
@@ -58,51 +63,38 @@ The three options are:
-
+-   [Tykky installation wrapper][contwrapper]
+
+     The Tykky installation wrapper is a piece of software developed at CSC
+     which wraps software installations inside an Apptainer/Singularity container
+     to improve startup times, reduce I/O load, and lessen the number of files on
+     large parallel file systems.
```
I kind of like the previous description because it lists some use cases, and doesn't mention containers. Containers sound scary to someone who has not been using them, and in Tykky case one doesn't need to care about the whole container thing to be able to use the tool accordingly

hdrei replied:
Ok, the previous comment is maybe unnecessary, since I see that the previous description was maybe misleading...

Chroxvi replied:
I have significantly reworked the software overview page. Does this new version address your concern?

hdrei commented on docs/software/index.md:
```diff
@@ -58,51 +63,38 @@ The three options are:
+installations in your scratch or flash directory is not recommended as these
+will be cleaned automatically in the future when the file system starts to fill
+up.
+
```
Not sure where to place this comment (because can't comment the unmodified part of the page), but below the 'Installing software' heading it says: "We offer three options to install software on LUMI", which is a bit conflicting taking account that on the 'Python packages' page there are described many methods more, although valid for certain use cases only.
In general, the section 'Python packages' should be mentioned on this page also elsewhere than in the warning box in the beginning of the page. For consistency of this section it would make sense to list it similarly than EasyBuild, Spack and Tykky

Chroxvi replied:
I have significantly reworked the software overview page. Does this new version address your concern?

hdrei replied:
Yes, it's much better now!

hdrei commented on docs/software/installing/container-wrapper.md:
```diff
@@ -1,43 +1,81 @@
 
-The tools provided by the container wrapper are accessible by loading the
+!!! warning "Tykky is NOT generally recommended for managing Conda/pip installations"
+    For some use cases, Tykky is an excellent tools for managing Conda/pip
```
```suggestion
    For some use cases, Tykky is an excellent tool for managing Conda/pip
```

hdrei reviewed (changes requested):
The main change suggestions that I see are:
 - The overview page could give a short description of the Python packages section also as "one way to install software" (or how to formulate it), to have the section structure consistent (as Python packages is placed under Installing software with EasyBuild, Spack and Tykky). 
 - The Python packages section: The listing of the ways to install could be somewhat clearer. Now one needs to use much extra focus to find the different possibilities to install (without reading all the text), or that's my experience at least. A suggestion: Change the headings from how recommended the methods are to the methods themselves? If the methods would be listed as headings, they would also appear on the upper right corner of the page as a list of contents

Chroxvi committed:
Apply straight forward suggestions from code review by hdrei

Co-authored-by: hdrei <78485804+hdrei@users.noreply.github.com>

Chroxvi committed:
Major rework of the software overview page.

Chroxvi committed:
Reworked the structure of the Python install overview as suggested by hdrei.

Chroxvi committed:
Minor updates to address review comments by hdrei.

Chroxvi commented:
@hdrei I have updated the PR to address your review comments. Please see check if all your concerns have been address.

- I have done a major rework of the software overview page. I like the new version a lot, but I am also a bit concerned about the overlap with the content in https://docs.lumi-supercomputer.eu/runjobs/lumi_env/Lmod_modules/ and https://docs.lumi-supercomputer.eu/runjobs/lumi_env/softwarestacks/.
- I have changed the structure of the Python packages installation page to more aggressively make use of sections as you suggested. I like the change. Initially, I also thought about creating a table as an alternative. However, I felt it would probably be difficult to make it easily comprehensible given the somewhat limited formatting options, markdown files provide.

Danzelot committed:
Fix link and show how to run container

hdrei commented on docs/software/containers/singularity.md:
```diff
@@ -58,42 +62,123 @@ where the command was run. Once the image has been pulled, the container can be
+    of the [Slingshot 11 interconnect][interconnect] when running MPI jobs, you
+    must make sure your packages are built against Cray MPICH. Cotainr does
+    **not** do any magic conversion of the packages specified in the Conda
+    environment to make sure they fit the hardware in LUMI. It simply installs
```
```suggestion
    environment to make sure they fit the hardware of LUMI. It simply installs
```

hdrei commented on docs/software/containers/singularity.md:
````diff
@@ -58,42 +62,123 @@ where the command was run. Once the image has been pulled, the container can be
+```
+
+where `my_conda_env.yml` is a file containing an [exported Conda
+environment][conda-env]. The resulting `my_container.sif` container may be run
+like any other [container job][container-jobs] on LUMI. For example:
+
+```bash
+$ singularity exec my_container.sif python3 --version
+```
````
The example given is not exactly "running a container job"? I'm not sure what/if to change here, and the link to LUMI examples in github kind of fixes this confusement

Chroxvi replied:
Maybe just change `--version` to `my_script.py` and mention that `my_script.py` is your Python script?

hdrei commented on docs/software/index.md:
```diff
@@ -10,38 +12,63 @@
-If you are looking for information on how to develop your own software on LUMI, consult the [developing
-section][developing-overview] instead. Please note that the LUMI user support team can only offer limited help with installing scientific software. This is further explained on the [Install policy](/software/policy/) page.
+On this page you find information about pre-installed software on LUMI as well
+guidance on ways to install additional software yourself. If you are
```
```suggestion
On this page you find information about pre-installed software on LUMI as well as guidance on ways to install additional software yourself. If you are
```

hdrei commented on docs/software/index.md:
```diff
@@ -10,38 +12,63 @@
+the [developing section][developing-overview] instead.
 
----
+Please note that the LUMI User Support Team can only offer limited help with
```
```suggestion
Please note that LUMI User Support Team can only offer limited help with
```

hdrei commented on docs/software/index.md:
```diff
@@ -10,38 +12,63 @@
-    based approach must be used. See below for the [container wrapper approach][contwrapper]
-    or [use of singularity][singularity-container].
+    based approach must be used. See the [Python packages installation guide
+    ][python-install] for en overview of options for installing Conda/pip
```
```suggestion
    ][python-install] for an overview of options for installing Conda/pip
```

hdrei commented on docs/software/index.md:
```diff
@@ -10,38 +12,63 @@
+    help on a specific module is through the `module help` command, but the LUMI
+    [Software Library pages][software-library] sometimes contain more information
+    on how to run the software, or more information about specific options that
+    were chosen when installing the software on LUMI.
```
```suggestion
    A full overview of software, that is either pre-installed in the [LUMI software
    stacks][software-stacks], or available as a user-installable LUMI-specific build recipe for the
    [EasyBuild][easybuild] package manager, can be found in the [LUMI software
    library][software-library]. The information, if a software is pre-installed or user-installable, can be found when you open the page of the specific software. For software installed in the
    [central LUMI software stacks][software-stacks], the first channel to get
    help on a specific module is through the `module help` command, but the LUMI
    [Software Library pages][software-library] sometimes contain more information
    on how to run the software, or more information about specific options that
    were chosen when installing the software on LUMI. Consult the [module environment page][module-env] for instructions on identifying the pre-installed software on LUMI through the module system instead.
    Read more about the user-installable software with EasyBuild in the [next section](#installing-additional-software). 
```

hdrei commented on docs/software/index.md:
```diff
@@ -10,38 +12,63 @@
+- Local software collections
 
-## Installing software
+    There is also pre-installed software available on LUMI provided by local LUMI
```
```suggestion
    There is also pre-installed software available provided by local LUMI
```

hdrei commented on docs/software/index.md:
```diff
@@ -10,38 +12,63 @@
+    on how to run the software, or more information about specific options that
+    were chosen when installing the software on LUMI.
 
+- Local software collections
```
```suggestion
- Software collections by local organizations
```

hdrei commented on docs/software/index.md:
```diff
@@ -50,59 +77,38 @@ The three options are:
     can be found in the [LUMI Software Library][software-library].
 
     Consult the [module environment page][module-env] for instructions on
```
```suggestion

```

hdrei commented on docs/software/index.md:
```diff
@@ -50,59 +77,38 @@ The three options are:
 
     Consult the [module environment page][module-env] for instructions on
-    identifying the already available software on LUMI through the module system instead. 
+    identifying the already available software on LUMI through the module system instead.
```
```suggestion

```

hdrei commented on docs/software/index.md:
```diff
@@ -50,59 +77,38 @@ The three options are:
     can be found in the [LUMI Software Library][software-library].
 
```
```suggestion
```

hdrei commented on docs/software/index.md:
```diff
@@ -50,59 +77,38 @@ The three options are:
     can be found in the [LUMI Software Library][software-library].
 
```
I'd like to replace the above lines with the following (but the website version doesn't let me). This adds the word "pre-installed", but is otherwise the same text:

An overview of (almost) all software that we provide pre-installed in the central software stack or for which we provide ready-to-use EasyBuild build recipes for a user installation can be found in the [LUMI Software Library][software-library].

Chroxvi replied:
I will push this change manually.

hdrei commented on docs/software/installing/python.md:
```diff
@@ -0,0 +1,147 @@
+
+Installing such a large number of small files to the user home folder or shared
+locations like `/scratch`, `/project`, or even `/flash`, and trying to load
+them from multiple processes at the same time, puts a lot of strain on the
```
```suggestion
them by multiple processes at the same time, puts a lot of strain on the
```

hdrei commented on docs/software/installing/python.md:
```diff
@@ -0,0 +1,147 @@
+Installing such a large number of small files to the user home folder or shared
+locations like `/scratch`, `/project`, or even `/flash`, and trying to load
+them from multiple processes at the same time, puts a lot of strain on the
+[Lustre file system][lustre] serving these storage locations. Lustre simple
```
```suggestion
[Lustre file system][lustre] serving these storage locations. Lustre simply
```

hdrei commented on docs/software/installing/python.md:
```diff
@@ -0,0 +1,147 @@
+!!! warning "The default Python is the OS Python"
+    When you log into LUMI, running `python3` without loading a module or using
+    a container will result in using the operating system Python installation.
+    This is an quite old Python installation (version 3.6) without any Scientific
```
```suggestion
    This is quite an old Python installation (version 3.6) without any Scientific
```

hdrei commented on docs/software/installing/python.md:
```diff
@@ -0,0 +1,147 @@
+
+In general, we recommend using [Singularity/Apptainer
+containers][singularity-containers] for managing Python installations instead
+of installing directly to the file systems. Using a container solves the "many
```
```suggestion
containers][singularity-containers] for managing Python installations. Using a container solves the "many
```

hdrei replied:
I'm trying to make the text sound like "the generally recommended installation methods" are not the only ones that are possible to use. (This becomes clear when one reads further to the next block "Installation methods recommended for specific use cases", or course.)

hdrei commented on docs/software/installing/python.md:
```diff
@@ -0,0 +1,147 @@
+containers][singularity-containers] for managing Python installations instead
+of installing directly to the file systems. Using a container solves the "many
+small files" performance problem and makes it easy to manage multiple different
+Python environments at the same time. You may either [use an existing
```
```suggestion
Python environments at the same time. To use a container, you may either [use an existing
```

hdrei commented on docs/software/installing/python.md:
```diff
@@ -0,0 +1,147 @@
+### Use the CSC software stack
+
+[CSC provides a software stack][csc-software-stack] on LUMI, similar to the
+software stacks provided on the Finish HPC systems, which contains some Python
```
```suggestion
software stacks provided on the Finnish HPC systems, which contains some Python
```

hdrei commented on docs/software/installing/python.md:
```diff
@@ -0,0 +1,147 @@
+
+### Use the CSC software stack
+
+[CSC provides a software stack][csc-software-stack] on LUMI, similar to the
```
```suggestion
[CSC provides a small additional software stack][csc-software-stack] on LUMI, similar to the
```

hdrei commented on docs/software/installing/container-wrapper.md:
```diff
@@ -1,43 +1,81 @@
+
+### Wrapping a basic Conda installation
+
+In order to wrap a basic [Conda][conda] installation, create an `<install_dir>`
```
```suggestion
In order to wrap a basic [Conda][conda] installation, create an installation directory
```

hdrei reviewed (changes requested):
This starts to look very good, I think. I made some small corrections/comments, and some small change suggestions

Chroxvi committed:
Apply suggestions from code review by hdrei

Co-authored-by: hdrei <78485804+hdrei@users.noreply.github.com>

Chroxvi committed:
Accept suggestion from review by hdrei

Co-authored-by: hdrei <78485804+hdrei@users.noreply.github.com>

Chroxvi committed:
A few changes to Python docs to address hdrei review comments.

Chroxvi commented:
@Nortamo, this PR makes some changes to the description of Tykky in the LUMI docs. Please chime in if you think some of it is inaccurate or otherwise needs to be changed.

Nortamo commented:
> @Nortamo, this PR makes some changes to the description of Tykky in the LUMI docs. Please chime in if you think some of it is inaccurate or otherwise needs to be changed.

Hi,
Sorry for the late reply I was away a week. 
Changes look good to me. New name is also sensible. Great work!

Chroxvi commented:
@Nortamo No problem! Thanks a lot for taking your time to look at this.

Chroxvi committed:
Reverted the name change for the LUMI container wrapper.
